src/librbd/internal.cc

```diff
@@ -1850,29 +1851,65 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       }
       assert(m_bl->length() == (size_t)r);
 
-      if (m_bl->is_zero()) {
+      if (!m_sparse_size) {
+	if (m_bl->is_zero()) {
+	  delete m_bl;
+	  m_throttle->end_op(r);
+	  return;
+	}
+
+	Context *ctx = new C_CopyWrite(m_throttle, m_bl);
+	auto comp = io::AioCompletion::create(ctx);
+
+	// coordinate through AIO WQ to ensure lock is acquired if needed
+	m_dest->io_work_queue->aio_write(comp, m_offset, m_bl->length(),
+					 std::move(*m_bl),
+                                         LIBRADOS_OP_FLAG_FADVISE_DONTNEED);
+      } else {
+        m_throttle->end_op(r);
```

Comment:
it seems C_Gather in Context.h is not exactly suitable here.  C_CopyWrite cannot become C_GatherSub roles in C_GatherBase. 

Here, actually,  we need to do completions for each sub_write, right? While in C_GatherBase, sub_finish() just --sub_existing_count, and does not call actual finisher callback here.